Ush 1193 - Filters should behave

ushahidlee opened the pull request:
**Problem**:

The filters being sent in requests to our backend API are inconsistent and in some cases nonsensical. Also, they often contain too much information, resulting in slower queries. 

- map view might request unstructured posts (which dont have any geo data),  (field: _include_unstructured_posts_)
- data view might request geolocated posts. (fields: _has_location_, _within_km_, _include_unmapped_)
- forms might include 0 (which represents "unknown form" on the frontend but has no meaning in the backend)
- arrays in the request might be included in the request even if empty.
- selected forms may alter when changing views


**Tickets**:

https://linear.app/ushahidi/issue/USH-1193/inconsistent-behavior-of-unknown-form-checkbox-and-include
https://linear.app/ushahidi/issue/USH-1194/unintended-visibility-of-unknown-form-checkbox-in-map-view
https://linear.app/ushahidi/issue/USH-1195/inconsistent-include-unstructured-posts-parameter-value-upon


**Solution**:

This PR attempts to move all the filter logic into the service, meaning any logical changes to filters are happening just before the request is sent. With all the logic in one place, it also attempts to remove meaningless fields from the filters.


**Testing**:

As this PR alters everything to do with filters, all filter behaviour should be observed while testing. However, the requests sent to the backend should be the specific focus of testing, with the following conditions:

All Requests:
- No empty arrays.
- _include_unstructured_posts_ should be present in every request.

Map View
- _include_unstructured_posts_ should always be false.
- _has_location_ should always be 'mapped'.

Data View
- _include_unstructured_posts_ should be true if "unknown form" is selected, false otherwise.
- _has_location_, _place_ and _within_km_ shouldnt be in the request.

In addition, each view has 2 requests associated with it. The above conditions should apply to both requests.

Map View
/api/v5/posts/stats
/api/v5/posts/geojson

Data View
/api/v5/posts/stats
/api/v5/posts

ushahidlee committed:
Bug 1 - Unknown Form on mapview

ushahidlee committed:
Bug 2 Fix : include_unstructured_posts in mapview

ushahidlee committed:
brackets, whats not to love?

ushahidlee committed:
Bug 3 - Filters.... looking good

ushahidlee committed:
Removed unstructured posts default filter value

ushahidlee committed:
added unstructured posts check just before endpoint

ushahidlee committed:
Tweaks for edge cases

ushahidlee committed:
more tweaks

ushahidlee committed:
Cleaned up more filter logic. Added more type safety.

ushahidlee committed:
a few tweaks

ushahidlee committed:
tweaks

ushahidlee committed:
Fixed edge case bug

Angamanga commented:
A few findings when testing this:
- The selectAll-button seems broken
- Sometimes, the result-numbers are 0, and sometimes they are not, for surveys:
- ![Screenshot 2024-04-10 at 09 38 55](https://github.com/ushahidi/platform-client-mzima/assets/8624777/d09e71ae-0687-45e8-9407-646da38245d7)
 It seems to happen after clicking "clear" and then "select all". Clearing the filters make it go back to normal again.
- When there is no data-sources connected to a deployment, there is still "Unknown form" in the filters-list. This is confusing to users and Cecilia has been asking about removing it but it may be out of scope for this ticket?
- All the rest seem to be working fine! 👍

ushahidlee committed:
Fixed bug around select all/clear

ushahidlee commented:
> A few findings when testing this:
> 
> * The selectAll-button seems broken

Fixed.

> * Sometimes, the result-numbers are 0, and sometimes they are not, for surveys:
> * ![Screenshot 2024-04-10 at 09 38 55](https://private-user-images.githubusercontent.com/8624777/321147573-d09e71ae-0687-45e8-9407-646da38245d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI4MzEwODEsIm5iZiI6MTcxMjgzMDc4MSwicGF0aCI6Ii84NjI0Nzc3LzMyMTE0NzU3My1kMDllNzFhZS0wNjg3LTQ1ZTgtOTQwNy02NDZkYTM4MjQ1ZDcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MTFUMTAxOTQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDYxZTJiOGVjM2Y3NDhmMDBhOWQxMzA3OTE4Y2QzZmQ2MWQ3YTEzYTZjMjBkYWY1ZjgzNzk3MmZmYTcwNGI5YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Ilmj0wdQs1BmDguoZAcb_1mz_EzRSaxqwiQfqFVLXMU)
>   It seems to happen after clicking "clear" and then "select all". Clearing the filters make it go back to normal again.

Fixed.

> * When there is no data-sources connected to a deployment, there is still "Unknown form" in the filters-list. This is confusing to users and Cecilia has been asking about removing it but it may be out of scope for this ticket?

So this is outside the scope of this ticket, but this PR does remove "Unknown Form" from the map view as it never makes sense there.

> * All the rest seem to be working fine! 👍

Angamanga reviewed (approved):
All fine now 🎉
